src/main/scala/strawman/collection/Iterator.scala

```diff
@@ -71,6 +72,186 @@ trait Iterator[+A] extends IterableOnce[A] { self =>
+     *  @return  The same iterator, and ''not'' a new iterator.
+     *  @note    This method mutates the iterator it is called on, which can be safely used afterwards.
+     *  @note    This method is mutually exclusive with `withPartial(true)`.
+     */
+    def withPadding(x: => B): this.type = {
+      pad = Some(() => x)
+      this
+    }
+    /** Public functions which can be used to configure the iterator before use.
+     *
+     *  Select whether the last segment may be returned with less than `size`
+     *  elements. If not, some elements of the original iterator may not be
+     *  returned at all.
+     *
+     *  @param x `true` if partial segments may be returned, `false` otherwise.
+     *  @return  The same iterator, and ''not'' a new iterator.
+     *  @note    This method mutates the iterator it is called on, which can be safely used afterwards.
+     *  @note    This method is mutually exclusive with `withPadding`.
+     */
+    def withPartial(x: Boolean): this.type = {
+      _partial = x
+      if (_partial == true) // reset pad since otherwise it will take precedence
+        pad = None
+
+      this
+    }
+
+    /** For reasons which remain to be determined, calling
+     *  self.take(n).toSeq cause an infinite loop, so we have
+     *  a slight variation on take for local usage.
+     *  NB: self.take.toSeq is slice.toStream, lazily built on self,
+     *  so a subsequent self.hasNext would not test self after the
+     *  group was consumed.
+     */
+    private def takeDestructively(size: Int): Seq[A] = {
+      val buf = new ArrayBuffer[A]
+      var i = 0
+      // The order of terms in the following condition is important
+      // here as self.hasNext could be blocking
+      while (i < size && self.hasNext) {
+        buf += self.next
+        i += 1
+      }
+      buf
+    }
+
+    private def padding(x: Int) = immutable.ImmutableArray.fill(x)(pad.get())
+    private def gap = (step - size) max 0
+
+    private def go(count: Int) = {
+      val prevSize = buffer.size
+      def isFirst = prevSize == 0
+      // If there is padding defined we insert it immediately
+      // so the rest of the code can be oblivious
+      val xs: Seq[B] = {
+        val res = takeDestructively(count)
+        // was: extra checks so we don't calculate length unless there's reason
+        // but since we took the group eagerly, just use the fast length
+        val shortBy = count - res.length
+        if (shortBy > 0 && pad.isDefined) res ++ padding(shortBy) else res
+      }
+      lazy val len = xs.length
+      lazy val incomplete = len < count
+
+      // if 0 elements are requested, or if the number of newly obtained
+      // elements is less than the gap between sequences, we are done.
+      def deliver(howMany: Int) = {
+        (howMany > 0 && (isFirst || len > gap)) && {
+          if (!isFirst)
+            buffer dropInPlace (step min prevSize)
```

Comment:
This code has been copied from the current collections, excepted that I replaced `trimStart` with `dropInPlace`.